error handling when path to dir provided instead of file

eliasyishak opened the pull request:
Fixes: #109450

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

eliasyishak committed:
validator created -- logic to be developed on windows

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
logic added for validator -- to be tested in windows env next

eliasyishak committed:
making call to built in windows 'systeminfo' from dart

eliasyishak committed:
Update windows_version_validator.dart

eliasyishak committed:
conditional to only run on windows

eliasyishak committed:
Update doctor.dart

eliasyishak committed:
error handling when path to dir provided instead of file

eliasyishak commented:
Need to create a test to be run in the device lab following instructions found here: https://github.com/flutter/flutter/blob/master/dev/devicelab/README.md

eliasyishak committed:
Merge pull request #1 from eliasyishak/master

merging version check

eliasyishak committed:
clean up

eliasyishak committed:
conditional update to check for version

eliasyishak committed:
conditional status info message added

eliasyishak committed:
refactoring to not use `dart:io`

eliasyishak committed:
changes requested during review

eliasyishak committed:
adding documentation for validator

eliasyishak committed:
fix to license file

eliasyishak committed:
fix to license file

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
Merge branch 'windows-version-check-in-doctor' of https://github.com/eliasyishak/flutter into windows-version-check-in-doctor

eliasyishak committed:
Merge branch 'windows-version-check-in-doctor' into master

eliasyishak committed:
Merge pull request #2 from eliasyishak/master

merging for latest code

eliasyishak committed:
fixing merging error

eliasyishak committed:
added 3 tests for each return case for windows version

- case 1: successful check (version 10 or greater)
- case 2: `systeminfo` process call fails
- case 3: version identified is less than 10

eliasyishak committed:
update to documentation + using async process run

eliasyishak committed:
simplified string parsing to utilize builtin regex functionality

eliasyishak committed:
remove todo

eliasyishak committed:
add optional indicator for new line chars

eliasyishak committed:
fixing regex pattern to look for start of line

eliasyishak committed:
added unit test for regex pattern validator

eliasyishak committed:
Merge branch 'flutter:master' into master

eliasyishak committed:
Merge pull request #3 from eliasyishak/windows-version-check-in-doctor

Windows version check in doctor

eliasyishak committed:
Merge pull request #4 from eliasyishak/master

updating feature branch with upstream master

eliasyishak committed:
removing ProcessException catch since doctor validator handles it already

eliasyishak committed:
Merge pull request #5 from eliasyishak/master

merging master into feature branch

eliasyishak committed:
Apply Jonah suggestions from code review

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

eliasyishak committed:
renaming variable

eliasyishak committed:
additional test added for a nonzero exit code from systeminfo

eliasyishak committed:
removing trivial static method, added constant for semver pattern

eliasyishak committed:
Merge branch 'master' of https://github.com/eliasyishak/flutter

eliasyishak committed:
Merge pull request #6 from eliasyishak/windows-version-check-in-doctor

Windows version check in doctor

eliasyishak committed:
Merge pull request #7 from eliasyishak/master

syncing master with feature branch

eliasyishak committed:
reverting try catch to find error upstream in function

eliasyishak committed:
injecting filesystem into screenshot command + conditionals added to validate filename and path

- updated screenshot_command_test.dart as well to inject MemoryFileSystem
- conditionals restrict users to alphanumeric characters, underscores, spaces, and hyphens for filename

eliasyishak commented on packages/flutter_tools/lib/src/commands/screenshot.dart:
```diff
@@ -123,16 +127,41 @@ class ScreenshotCommand extends FlutterCommand {
+      throwToolExit(
+        'The provided filename is invalid\n'
+        'Allowed filename characters: "a-z", "A-Z", "0-9", "_", "-"\n'
+        'Allowed extensions: jpeg, jpg, png)\n'
```
Might want to revisit this to determine what file types are supported through all targets

eliasyishak committed:
refactoring to make outputFile validation a static method

Makes it easier to unit test

eliasyishak committed:
included tests for validating screenshot output file

jonahwilliams commented:
I don't think this is the right approach: in general this feels far too restrictive, especially as it is perfectly reasonable for folks to give screenshots whatever extension they want or use non-ascii characters.


Is there an exception thrown if we create a file with an unsupported name, like "."? If so, we could catch that exception instead. But consider shifting from an allow based approach to a deny based approach

eliasyishak commented:
Yeah it definitely was a more restrictive approach, and that's understandable if we would like to make it less restrictive. We can remove any character restriction and file type restriction on the pattern and just make sure the directory is valid?

Because the tool does actually allow for files that don't even have any extensions, ie. `-o image` will produce a file for the user. Does that approach make sense?

christopherfujino commented:
> Yeah it definitely was a more restrictive approach, and that's understandable if we would like to make it less restrictive. We can remove any character restriction and file type restriction on the pattern and just make sure the directory is valid?
> 
> Because the tool does actually allow for files that don't even have any extensions, ie. `-o image` will produce a file for the user. Does that approach make sense?

isn't it reading the extension and then determining what encoding to use based on that? so if you pass `-o image`, what kind of image is it creating? is there a default extension, or is it just not doing anything and exiting 0?

jonahwilliams commented:
As far as I can tell, the tool does not use the file extension to determine what kind of file output.

christopherfujino commented:
> As far as I can tell, the tool does not use the file extension to determine what kind of file output.

this is simctl

eliasyishak commented:
I'm not sure how simctl handles it but I have seen that the file gets generated, although I wasn't able to open it on my mac. Which was why I initially thought to add the extension requirement and possibly expand to different file types as needed.

jonahwilliams commented:
but simctl isn't the only backend here, and based on the snapshot type argument we can also create a "skia" and "rasterizer" type

eliasyishak committed:
removed restrictive conditional on filename – kept dir check

- added additional transformation for outputFile incase users pass a trailing directory delimiter (ie. '-o image.png/' will result in a file being output at $PWD/image.png

eliasyishak committed:
remove path import

eliasyishak commented on packages/flutter_tools/lib/src/commands/screenshot.dart:
```diff
@@ -183,6 +191,37 @@ class ScreenshotCommand extends FlutterCommand {
+  /// return a 0 exit code even if outputFile is not valid
+  static File validateOutputFile(File outputFile, FileSystem fs) {
+
+    // Will make sure that if the user passes '-o image.png/' the file
```
@jonahwilliams i added this additional functionality to catch any errors if the user provides a trailing delimiter. Do you think it's worth keeping this or should the tool throw an exception?

jonahwilliams replied:
I would not modify the file the user provides to this script, either accept it if it is valid or deny if it isn't. Consider the case where someone uses this command in part of an automated test script, and said test script has a bug that gets an invalid path. You would want that to fail as soon as possible, instead of accidentally passing for some time, or failing because a different command that accepts a file didn't make the same change to the file name to make it valid

eliasyishak replied:
Great point, I'll make the adjustments

eliasyishak committed:
removing functionality that allows trailing path separators

jonahwilliams commented on packages/flutter_tools/lib/src/commands/screenshot.dart:
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
+    const Platform platform = LocalPlatform();
+
+    // Conditional to check for trailing path separator
+    if (outputFile.path.endsWith(platform.pathSeparator)) {
```
This is available on `path`.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/screenshot.dart:
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
+    // Conditional to check for trailing path separator
+    if (outputFile.path.endsWith(platform.pathSeparator)) {
+      throwToolExit(
+        'The provided path cannot end with a path separator\n'
+        'Path provided: "${outputFile.path}"'
```
I would rephrase this to indicate that we expect a file path but got a directory path.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/screenshot.dart:
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
+    }
+
+    // Conditional for validating directory is valid
+    if (!fs.directory(outputFile.dirname).existsSync()) {
+      throwToolExit(
+        'The provided path to file needs to have a directory that exists\n'
+        'Directory: "${outputFile.dirname}" does not exist'
+      );
+    }
```
Use `fs.file(outputFile).parent.existsSync()` - I'm not sure if dirname includes the contents of absolute paths.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/screenshot.dart:
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
+    }
+
+    // Conditional for validating filename
+    if (outputFile.basename.isEmpty || outputFile.basename == '.') {
```
what about `..` or `../`?

eliasyishak committed:
simplifying logic to just check for a valid output after screenshot command

eliasyishak committed:
remove unused import

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/screenshot_command_test.dart:
```diff
@@ -26,42 +27,89 @@ void main() {
     });
   });
+
+  group('Screenshot file validation', () {
```
Since all this command does now is check if the file exists, you only need a test case for 1) file exists and 2) file doesn't exist

eliasyishak committed:
simplifying tests to successful and fail results

jonahwilliams reviewed (approved):
LGTM
